Support for seppuku flight controller

mlyle opened the pull request:
Seppuku is a new flight controller with full sensor suite on SPI, STM32F405, and graphical OSD.

@dtfuhf will be producing it; I have designed the hardware but have no financial interest in its success or merge.  Note that this is not formally announced and we are not prepared to talk about final pricing, availability dates, features, etc.

Closed source currently.  Schematics will open 3mos after first customer shipment.

I'd like to defer many of these to after this PR.. but tracking the list here for now:

* [ ] Panic lights to ws2811!
* [ ] Drive mask in bootloader
* [x] Final graphics in GCS
* [x] Wizard support in GCS
* [x] Note rx jumper on GCS board config page
* [x] Further investigation bmp280 oddness
* [x] Get proper OSD levels by default
* [x] ~~Anomaly: OSD doesn't work every other boot~~  Seems to be @dtfuhf's LCD display 
* [x] ~~Anomaly: black/white level drift (probably small passive value change)~~ Seen only on one unit
* [ ] ~~Share SVG between board config and connection diagram~~

mlyle committed:
flight: seppuku: add target as a copy of quanton

mlyle committed:
seppuku: make initial target makefile edits

mlyle committed:
seppuku: add initial hwseppuku XML

mlyle committed:
seppuku: remove gdb files

mlyle committed:
seppuku: set to generic firmware product ID

mlyle committed:
seppuku: adjust docstrings for target

mlyle committed:
seppuku: hack pios_board into approximate shape

mlyle committed:
seppuku: further work on board defs

mlyle committed:
seppuku: complete board_hw_defs except video

mlyle committed:
seppuku: preliminary bmi160 gyro support

mlyle committed:
seppuku: get osd & pios_video building, configured

mlyle committed:
seppuku: attempt to generate black/white levels

tracernz committed:
GCS:Seppuku: Initial support

tracernz committed:
GCS: Rework connection diagrams, add to hardware page

tracernz committed:
GCS:Seppuku: Initial graphics

mlyle committed:
seppuku: support for lis3mdl onboard mag.

Also move sensor configurartions to board_hw_defs.

mlyle committed:
seppuku: pick appropriate PLL factors

mlyle committed:
seppuku: fixes to compile with current code

mlyle committed:
seppuku: adjust annunciator indices

mlyle committed:
seppuku: correct some usb strings

mlyle committed:
seppuku: correct settings sector size

mlyle committed:
seppuku: only use vsense in loader

Necessary because we use it to drive an inverter later.

mlyle committed:
seppuku: temp heartbeat LED: invert sense

mlyle committed:
seppuku: correct OSCILLATOR_FREQ

mlyle committed:
seppuku: vsense: correct to have pulldown

Might be nice to drive this downwards briefly before sampling, too.

mlyle committed:
seppuku: provide minimal feedback if clock doesn't start

mlyle committed:
seppuku: small cleanups/fixes

mlyle committed:
seppuku: make consistent with revolution target

mlyle committed:
seppuku: use BMP280

mlyle committed:
seppuku: set buzzer pin, fix PPM

mlyle committed:
seppuku: bl: add ws2811 support

mlyle committed:
seppuku: fix up for autotune partition

Also slightly rearrange partition map to free up the desirable 64k
sector for future use.

mlyle committed:
seppuku: adjust for rev B

mlyle committed:
pios_video: changes for seppuku

1. pios_video: teach about TIM9
2. pios_video: properly treat TIM9 as double speed
2. pios_video: allow hsync to use filtered timer input 1

mlyle committed:
seppuku: attempt to set the proper sensor orientation

mlyle committed:
osd: use better logo on seppuku, print target name

mlyle commented:
jenkins, build artfacts please

mlyle commented:
jenkins, build artifacts please

mlyle commented:
Just a little more background info:
![pinmap 1](https://cloud.githubusercontent.com/assets/90903/21035226/678bfff4-bd72-11e6-882e-30642c14bf08.png)

mlyle committed:
seppuku: turn on floating point printf support

mlyle committed:
seppuku: adjust PWM levels

Increase the PWM frequency a little, and get the levels to line up
with Brain's existing values in the uavos.

(20+m)/x = (120/255), 110+m/x = (240/255), m=70, x~=192

mlyle committed:
seppuku: drive mask/level signals harder

yamahabest commented on ground/gcs/src/plugins/boards_dronin/seppuku.cpp:
```diff
@@ -0,0 +1,198 @@
+bool Seppuku::isInputConfigurationSupported(InputType type)
+{
+    Q_UNUSED(type)
+    return false;
```
Shouldn't this "false" be a "true"?
Else you wouldn't be needing the set- and getInputType functions.

yamahabest replied:
And PWM should probably be added as not supported?

mlyle replied:
Yah, it's marked that way for now because it's definitely not correct.  Adding to the list of further work needed.

mlyle committed:
systemmod: attempt to prevent brief blink of other color

droney commented:
hello icee - i look forward to a reasonably priced replacement for the original brain. Not sure if you intend to support an external mag (i2c) interface. The original brain failed to include pull ups on the SCL/SDA signals, so external mag was unreliable and required the resistors to be added externally.

mlyle committed:
seppuku: rtc: adjust frequency

mlyle commented:
@droney I want to hold off talking too much about things until the real announcement.  But one of the serials (USART3) can be an I2C and has 2.2K pullups to VCC, so it ought to be comfortable running at 400KHz.

mlyle committed:
seppuku: fix off-by-10 ADC channel mapping :P

mlyle committed:
seppuku: enable smartaudio (in lame / usart way)

mlyle committed:
pios_video: add support for falling-edge hsync

mlyle committed:
seppuku: enable brownout detector on first fw boot

Slow rising power definitely made sepp hang on start before.
This fixes it.

mlyle committed:
seppuku: slow down some edge rates

mlyle committed:
seppuku: support usart6

mlyle committed:
pios_video: synchronize in a more robust way

Hopefully removes NTSC related glitches on Seppuku.

Adjust hsync appropriately, and allow target-specific hsync offsets.

Allow broader user-hsync offset.

mlyle committed:
brainre1: support get num leds call

mlyle commented:
Rebased.  Confirmed OK OSD on OG brain on PAL after changes.

jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

mlyle committed:
seppuku: correct magnetometer orientation

mlyle commented:
jenkins, build artifacts please

mlyle committed:
pios_spi: fixes

- Don't change AF on SSEL pins unless it matters.
- Use correct PIOS_Semaphore constant for infinite blocking

mlyle committed:
seppuku: small changes to interrupt line pulling

mlyle committed:
seppuku: introduce a hack to prevent DMA

Video DMA (APB) and WS2811 DMA (AHB) conflict on DMA2, causing errors on
secondary sensors.  Relates to ST errata 2.1.10 in
http://www.st.com/content/ccc/resource/technical/document/errata_sheet/0a/98/58/84/86/b6/47/a2/DM00037591.pdf/files/DM00037591.pdf/jcr:content/translations/en.DM00037591.pdf

mlyle commented on flight/Modules/OnScreenDisplay/onscreendisplay.c:
```diff
@@ -135,6 +135,8 @@ const char IMPERIAL_DIST_UNIT_LONG[] = "M";
 const char IMPERIAL_DIST_UNIT_SHORT[] = "ft";
 const char IMPERIAL_SPEED_UNIT[] = "MPH";
 
+bool video_active;
```
@mluessi this is much closer to the "real fix"

mlyle commented:
jenkins, build artifacts please

mlyle committed:
seppuku: correct uart 4 config

mlyle commented:
jenkins, build artifacts please

tracernz committed:
GCS:UAVOWidgetUtils: Add widget helper functions

Before disabling widgets with UAVO relations from outside ConfigTaskWidget was problematic because ConfigTaskWidget would ignore what you had done and vice-versa. ConfigTaskWidget::setWidgetEnabled to the rescue.

Setting dynamic properties on widgets that are used in the widget stylesheet is problematic because they stylesheet isn't re-evaluated automatically. This provides a small helper function to handle this rather than re-writing it dozens of times elsewhere.

tracernz committed:
GCS:Seppuku: Add configuration UI

tracernz committed:
GCS:Seppuku: Remove incorrect DSM bind warnings

tracernz commented:
jenkins, build artifacts please

tracernz committed:
GCS:Seppuku: Add I2C pin markings on board image (dynamic)

tracernz committed:
GCS:Seppuku: Add receiver voltage jumper reminder

mlyle committed:
osd: Don't average altitudes right at the beginning

The baro may be pretty far off at that point.

mlyle committed:
seppuku: attempt to update config/connection for final art

tracernz commented on ground/gcs/src/plugins/boards_dronin/seppuku.ui:
```diff
@@ -576,6 +576,22 @@ margin: 2px 0 2px;</string>
                 </property>
                </widget>
               </item>
+              <item row="1" column="0" colspan="2">
```
Ahh, have this done (and dynamic so it goes away when you have receiver disabled. Few other commits at home that need pushed too IIRC but I didn't here back from you the other day. ;)

tracernz committed:
GCS:Seppuku: Fix board element ID

tracernz commented:
@mlyle Fixed the element ID on the board image so it displays. Needs replaced with a transparent version too. Also possibly alignment issues with text but Inkscape doesn't work properly on macOS and Photoshop/Creative Cloud wants to use all my mobile data to unf--- itself and let me use PS. :(

tracernz commented:
@mlyle Also, you deleted the board image used in the uploader? https://github.com/d-ronin/dRonin/pull/1480/files#diff-5128b30b6559dd3c0c1ff5d23d9a000bR183
```
[Warning] QPixmap::scaled: Pixmap is a null pixmap
```

mlyle commented:
Ah, that's the uploader one.

I deleted it because it was attempted to be referenced from the hardware diagram but it pointed to a file in your home directory, and I thought that was the only use.

I'll look at pictures/art assets soon.

tracernz commented:
Haha. I'll check it all out when I'm home tomorrow evening. :-)

tracernz committed:
GCS:Seppuku:Minor improvements to configuration UI

tracernz commented:
Connection diagram WIP 😀

pug398 commented:
jenkins, build artifacts please

tracernz committed:
GCS:Seppuku:Add new board image

tracernz committed:
GCS:UAVOField:Add helper to get element name

tracernz committed:
GCS:Connections:Use UAVOs to draw diagrams

Also, as part of fixing revo/cc3d, finally split them, correct the PPM pin issue, and put SBUS on the correct port! Fixes #545 properly.

tracernz committed:
GCS:Seppuku: Add QuadX, all receivers, OSD to connection diagram

tracernz committed:
GCS:Seppuku:Support setup wizard

tracernz committed:
UAVObjectField: Accept int as value for enum

To allow @mlyle to do what he did, and for general niceness. :)

tracernz committed:
GCS:Seppuku:Alignment fix in connection diagram

tracernz commented:
Connection diagram has everything on Rx port, only quad x vehicle type, OSD (cam/vtx), and external I2C. Should be enough for now.

jenkins, build artifacts please

DTFUHF commented:
Nice, can't wait to try this target.

On Fri, Jan 13, 2017, 02:07 Michael Corcoran <notifications@github.com>
wrote:

> Merged #1480 <https://github.com/d-ronin/dRonin/pull/1480>.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/d-ronin/dRonin/pull/1480#event-921855555>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEitHqerSjRxoPPAhBIiFQjxAEfjux1mks5rRyKpgaJpZM4LIgYn>
> .
>

LitterBugs commented:
Pulling latest to give it a try.  Had issue with compass being off 180 degrees yesterday, but may have old version.

pug398 commented:
@LitterBugs The compass should be right using the latest. After a 6 point calibration on the internal mag could you note the mag values with it due north and let me know? 

Also under settings>sensorsettings>MagBias what bias is shown particularly the z bias?

mlyle commented:
@LitterBugs re: latest -- I suggest a current nightly build-- http://jar.lyle.org:8080/job/dronin/838/ for instance.  Anything on jar.lyle.org:8080 made at 5:03 is a nightly that contains all changes in the development release.

LitterBugs commented:
That's really odd.... somehow I had downgraded to and older version that still had issues with NTSC last week.  Must have deleted the wrong directory tree when messing around with the TriFlight version on my F3FC.   Ah well, back on the right track with 838.

pug398 commented:
As far as the picture quality NTSC is pretty good now. Depending on camera may need to lower whitelevel to lock sync in or raise blacklevel to stop the white from streaking across but pretty decent picture overall. I had to lower camera voltage to 5v (from 12v) because it was overdriving video input.

LitterBugs commented:
Yeah, it was completely my fault.  Had downgraded to one of the first releases last week when I thought I was upgrading.  Working on reconfiguring with 838 now.  Too many pots on the stove....

LitterBugs commented:
Did find that the optimal black level needed to be raised from 20 to 35 with my current camera/VTX.  My ground display worked fine, but my DVR and Goggles would loose sync under some lighting conditions otherwise.  @pug398, Do you still want my compass info since it calibrated ok?

pug398 commented:
I guess my main question is under settings>sensorsettings>MagBias is z around -10,000?

LitterBugs commented:
X = 11014.64160156
Y = 3280.28002930
Z = -23200.98242188

pug398 commented:
Obviously in this application mag bias can run in the tens of thousands. I had not seen any go that high before. PH seem to hold well in the wind so I didn't really think it was a problem but more of a curiosity. Thanks for checking it for me.

LitterBugs commented:
Tried out 838 at the field tonight.  Is the Baro supposed to reset at arm?  Seemed to be set to a previously stored value as can be seen in the video linked below.   Compass worked great, OSD worked great, Controller was flying great.  Got a bit greedy on a dive and put it in the dirt.
https://youtu.be/FFZ2ejIWuk0

mlyle commented:
@litterbugs -- is there any chance you have on the OSD page AltitudeScaleSource set to Nav?  Right now I think that'll end up reporting altitude over sea level (well, pressure altitude, which on a normal day will be altitude over sea level).

LitterBugs commented:
It is set to "Baro".  I am at roughly 980' above sea level.  It was at 0 last night after doing the 6 point calibration.  

Edit: No GPS connected either.

LitterBugs commented:
Hi  Michael,
Just wanted to give you a heads up that I ran into reboot issues today 
with the "Test" Seppuku.  When the sun is in the upper right corner of 
the video, you can see some really bad video artifacting, and eventially 
it caused the OSD to dissappear and then the FC to reboot.  Uploading 
video and will post up on dRonin board.

Regards,
LitterBug
